src/NuGet.Services.AzureSearch/Catalog2AzureSearch/CatalogLeafFetcher.cs

```diff
@@ -0,0 +1,253 @@
+                    continue;
+                }
+
+                if (available.ContainsKey(version))
+                {
+                    _logger.LogDebug(
+                        "For {PackageId}, version {Version} was already discovered to be available.",
+                        packageId,
+                        version);
+                    continue;
+                }
+
+                _logger.LogInformation(
+                    "Looking for the catalog leaf for {PackageId} {Version}.",
+                    packageId,
+                    version);
+
+                var info = GetPageInfo(pageUrlToInfo, version);
+                if (info == null)
+                {
+                    _logger.LogWarning(
+                        "No page was found for {PackageId} {Version}. Page ranges were: {Ranges}",
+                        packageId,
+                        version,
+                        ranges);
+                    unavailable.Add(version);
+                    continue;
+                }
+
+                if (info.VersionToItem == null)
+                {
+                    _logger.LogInformation(
+                        "Fetching the items for page {PageUrl}. Range: {Range}",
+                        info.Page.Url,
+                        info.RangeString);
+
+                    var page = await _registrationClient.GetPageAsync(info.Page.Url);
+                    info.SetVersionToItem(page.Items);
+                }
+
+                if (!info.VersionToItem.TryGetValue(version, out var item))
+                {
+                    _logger.LogWarning(
+                        "No registration leaf item found for {PackageId} {Version} on {PageUrl}",
+                        packageId,
+                        version,
+                        info.Page.Url);
+                    unavailable.Add(version);
+                    continue;
+                }
+
+                _logger.LogInformation(
+                    "Fetching the catalog leaf for {PackageId} {Version} from {LeafUrl}",
+                    packageId,
+                    version,
+                    item.CatalogEntry.Url);
+
+                var leaf = await _catalogClient.GetPackageDetailsLeafAsync(item.CatalogEntry.Url);
+                available[version] = leaf;
+
+                if (leaf.IsListed())
+                {
+                    _logger.LogInformation(
+                        "{PackageId} {Version} was found to be listed. Metadata from {Url} will be used.",
+                        packageId,
+                        version,
+                        leaf.Url);
+                    return;
+                }
+                else
+                {
+                    _logger.LogInformation(
+                        "{PackageId} {Version} was found to be unlisted. This will not be used as a latest version.",
+                        packageId,
+                        version,
+                        leaf.Url);
```

Comment:
>leaf.Url [](start = 24, length = 8)

This is missing a field in the log message.

Reply:
Nice catch. I guess Microsoft.Extensions.Logging is okay with this (no throw).